torch/csrc/jit/tensorexpr/kernel.cpp

```diff
@@ -1184,7 +1243,12 @@ Tensor* TensorExprKernel::computeValue(const torch::jit::Value* v) {
 
     case aten::abs: {
       return computeOneOperand(
-          "aten_abs", v, [](const ExprHandle& a) { return fabs(a); });
+          "aten_abs",
+          v,
+          [](const ExprHandle& a) {
+            return fabs(promoteHalfAndIntegerToFloat(a));
```

Comment:
Is promotion to float necessary for cuda?  I guess it probably doesn't hurt anything...

Reply:
You mean necessary for the LLVM back-end? Because I think it'd work in CUDA through implicit promotion in the generated CUDA C++, but it wouldn't in LLVM IR which doesn't have implicit promotions.

Reply:
I will check in a fix so that integral types do not require promotion to float for `abs`. 

However, promotion from half to float is probably still needed. Since f16 is not a native type for normal CPU backend (or `std::xxx` functions), unless we develop math libraries for `f16`, promoting to f32 would be the easiest implementation. But for devices w/ native support for `f16` (e.g., gpu and most accelerators), this promotion may kill the performance benefit of `f16`. 

BTW, when we lower from `torch::jit::Node` to `IntrinsicsOp` in `TensorExprKernel::compile()`, we are not supposed to be platform-specific, so we have to normalize to the safest implementation (that works for all platforms). 

I feel the right approach is to make the code-gen of IntrinsicsOps to handle Half to float promotion. NOTE: as eager runtime supports more fp16 in kernel functions, this issue will pop up more. (If so, can we do this right approach in a separate issue to limit the scope of this one?)

Reply:
> can we do this right approach in a separate issue to limit the scope of this one

Yeah I think it's totally fine to do the "right approach" in future diffs.

Reply:
Last time handling `Half` in kernel.cpp was brought up the consensus was it was better to do it later in backend specific logic. 

Could IREval just call the Half-Rewriter before attempting to execute to handle Half ? Similar to how the other backends handle half:  https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/tensorexpr/llvm_codegen.cpp#L317

Reply:
Opened issue #48864 to track this enhancement. Thanks @eellison for the pointer to half-rewriter, I think letting codegen handles more types is the right approach.